### PR TITLE
[RayCluster][CI] add e2e tests for RayClusterStatusCondition

### PR DESCRIPTION
**Pull request by rueian**
## Why are these changes needed?

Continue https://github.com/ray-project/kuberay/issues/2645.

This PR adds e2e tests for the `HeadPodReady`, `RayClusterProvisioned`, and `ReplicaFailure` status conditions.

## Checks

- [x] I've made sure the tests are passing.
- Testing Strategy
   - [x] Unit tests
   - [x] Manual tests

### CODE REVIEW

File: ray-operator/config/manager/manager.yaml [hunk 1]
```diff
@@ -26,7 +26,8 @@ spec:
       containers:
       - command:
         - /manager
-#        args:
+        args:
+        - --feature-gates=RayClusterStatusConditions=true
```
Comment (rueian) [inline ray-operator/config/manager/manager.yaml:30]: We need this to make [buildkite/ray-ecosystem-ci-kuberay-ci/test-sample-yamls-nightly-operator](https://buildkite.com/ray-project/ray-ecosystem-ci-kuberay-ci/builds/5712#0193d390-b9f9-42ec-9c81-1754786d87a1), where the feature gate is not enabled by default, to pass.

**Comment by rueian** [comment]
Hi @andrewsykim, please take a look. Thanks.

File: ray-operator/config/manager/manager.yaml (hunk 1 above)
Comment (andrewsykim) [reply ray-operator/config/manager/manager.yaml:30]: Isn't that using nightly build though? Nightly build should have this feature gate enabled by default at this point?

Comment (rueian) [reply ray-operator/config/manager/manager.yaml:30]: Oh, I copied the wrong link. I should mean this one [buildkite/ray-ecosystem-ci-kuberay-ci/test-sample-yamls-latest-release](https://buildkite.com/ray-project/ray-ecosystem-ci-kuberay-ci/builds/5712#0193d390-b9fa-4fad-9911-cbf9298d1794)

Comment (rueian) [reply ray-operator/config/manager/manager.yaml]: > ah I see, makes sense. Can you add a comment that we can remove this once v1.3 is released?

Yes. The comment is added.

File: ray-operator/test/support/yaml.go [hunk 2]
```diff
@@ -55,3 +55,19 @@ func KubectlApplyYAML(t Test, filename string, namespace string) {
 	assert.NoError(t.T(), err)
 	t.T().Logf("Successfully applied %s", filename)
 }
+
+func KubectlApplyQuota(t Test, namespace, quota string) {
+	t.T().Helper()
+	kubectlCmd := exec.CommandContext(t.Ctx(), "kubectl", "create", "quota", namespace, "-n", namespace, quota)
+	err := kubectlCmd.Run()
+	assert.NoError(t.T(), err)
```
Comment (kevin85421) [inline ray-operator/test/support/yaml.go:63]: `assert.NoError` does not fail immediately; the error is only printed after the test finishes and `"Successfully applied quota %s in %s"` will always be printed no matter the command succeed or not. Should we make it fail fast?

File: ray-operator/test/support/yaml.go [hunk 3]
```diff
@@ -55,3 +55,19 @@ func KubectlApplyYAML(t Test, filename string, namespace string) {
...
+	t.T().Logf("Successfully applied quota %s in %s", quota, namespace)
+}
+
+func KubectlDeleteAllPods(t Test, namespace string) {
+	t.T().Helper()
+	kubectlCmd := exec.CommandContext(t.Ctx(), "kubectl", "delete", "--all", "pods", "-n", namespace)
+	err := kubectlCmd.Run()
+	assert.NoError(t.T(), err)
```
Comment (kevin85421) [inline ray-operator/test/support/yaml.go:71]: same

**Review by kevin85421 — COMMENTED** [review]
Nice

File: ray-operator/test/support/yaml.go (hunk 3 above)
Comment (rueian) [reply ray-operator/test/support/yaml.go]: Good catch! I replaced all `assert` with `require` which will fail fast in the `yaml.go file`. I can file a PR to replace all `assert` usages with `require` throughout the entire project if needed.

**Comment by kevin85421** [comment]
can you fix the conflict?

**Comment by rueian** [comment]
> can you fix the conflict?

The conflict came from https://github.com/ray-project/kuberay/pull/2660/files#r1887796931 where the `acquire` package dependency was removed. Right now the conflict is resolved by using the `require` package, but I think relying only on the standard library is also a good idea. Do we now prefer not to use any `acquire` or `require` packages? If yes, I can open another PR to remove them.

**Comment by kevin85421** [comment]
> Right now the conflict is resolved by using the require package, but I think relying only on the standard library is also a good idea. Do we now prefer not to use any acquire or require packages

I don't have any preference.

File: ray-operator/test/support/yaml.go [hunk 4]
```diff
@@ -28,32 +28,46 @@ func DeserializeRayClusterYAML(t Test, filename string) *rayv1.RayCluster {
...
 }
 
 func KubectlApplyYAML(t Test, filename string, namespace string) {
 	t.T().Helper()
 	kubectlCmd := exec.CommandContext(t.Ctx(), "kubectl", "apply", "-f", filename, "-n", namespace)
 	err := kubectlCmd.Run()
-	if err != nil {
-		t.T().Fatalf("Failed to apply %s to namespace %s: %v", filename, namespace, err)
-	}
+	require.NoError(t.T(), err)
```
Comment (kevin85421) [inline ray-operator/test/support/yaml.go:55]: I think `err` is not very informative. From what I recall, it seems like just something similar to "exit 1".

Comment (rueian) [reply ray-operator/test/support/yaml.go]: Updated. Manual error messages have been added to `require.NoError()` calls.

**Review by kevin85421 — APPROVED** [review]

**Pull request merged**